File: java/src/main/java/io/binac/leetcode/FirstMissingPositive.java
package io.binac.leetcode;

/**
 * Given an unsorted integer array, find the smallest missing positive integer.
 * <p>
 * <p>Example 1:
 * <blockquote><pre>
 *     Input: [1,2,0]
 *     Output: 3
 * </blockquote></pre>
 * Example 2:
 * <blockquote><pre>
 *     Input: [3,4,-1,1]
 *     Output: 2
 * </blockquote></pre>
 * Example 3:
 * <blockquote><pre>
 *     Input: [7,8,9,11,12]
 *     Output: 1
 * </blockquote></pre>
 * Note:
 * <p>
 * <p>Your algorithm should run in O(n) time and uses constant extra space.
 */
public class FirstMissingPositive {
    public static class Solution1 {
        private static void swap(int[] x, int a, int b) {
            int t = x[a];
            x[a] = x[b];
            x[b] = t;
        }

        public int firstMissingPositive(int[] nums) {
            if (nums.length == 0)
                return 1;

            for (int i = 0; i < nums.length; ++i) {
                while (1 <= nums[i] && nums[i] <= nums.length && nums[nums[i] - 1] != nums[i])
                    swap(nums, i, nums[i] - 1);
            }
            for (int i = 0; i < nums.length; ++i) {
                if (nums[i] != i + 1)
                    return i + 1;
            }
            return nums.length + 1;
        }
    }
}

File: python/leetcodepy/first_missing_positive.py
"""
Given an unsorted integer array, find the smallest missing positive integer.

Example 1:

Input: [1,2,0]
Output: 3

Example 2:

Input: [3,4,-1,1]
Output: 2

Example 3:

Input: [7,8,9,11,12]
Output: 1

Note:

Your algorithm should run in O(n) time and uses constant extra space.
"""
from typing import List


class Solution1:
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        size = len(nums)
        for i in range(size):
            while 0 < nums[i] <= size and nums[i] != i + 1 and nums[nums[i] - 1] != nums[i]:
                self.swap(nums, i, nums[i] - 1)
        for i in range(size):
            if nums[i] != i + 1:
                return i + 1
        return size + 1

    @staticmethod
    def swap(nums: List[int], i: int, j: int):
        nums[i], nums[j] = nums[j], nums[i]
